File: tests/scripts/test-ref-configs.pl
#!/usr/bin/env perl

# test-ref-configs.pl
#
# Copyright The Mbed TLS Contributors
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

use warnings;
use strict;
use Getopt::Std;
use POSIX qw(SIGINT);

my %configs = (
    'config-ccm-psk-tls1_2.h' => {
        'compat' => '-m tls12 -f \'^TLS-PSK-WITH-AES-...-CCM-8\'',
        'test_again_with_use_psa' => 1
    },
    'config-ccm-psk-dtls1_2.h' => {
        'compat' => '-m dtls12 -f \'^TLS-PSK-WITH-AES-...-CCM-8\'',
        'opt' => ' ',
        'opt_needs_debug' => 1,
        'test_again_with_use_psa' => 1
    },
    'config-no-entropy.h' => {
    },
    'config-suite-b.h' => {
        'compat' => "-m tls12 -f 'ECDHE-ECDSA.*AES.*GCM' -p mbedTLS",
        'test_again_with_use_psa' => 1,
        'opt' => ' ',
        'opt_needs_debug' => 1,
    },
    'config-symmetric-only.h' => {
        'test_again_with_use_psa' => 0, # Uses PSA by default, no need to test it twice
    },
    'config-thread.h' => {
        'opt' => '-f ECJPAKE.*nolog',
        'test_again_with_use_psa' => 1,
    },
);

sub HELP_MESSAGE {
    my ($fh) = @_;
    print $fh <<EOF;
Usage: $0 [OPTION]... [CONFIG_NAME [...]]

For each reference configuration file in the configs directory, build the
configuration, run the test suites and, for some configurations, TLS tests.
If given one or more config name, only test these configurations.

Options:
  -k            Keep going after errors
  --help        Print this help and exit

Config names:
EOF
    print $fh map {"  $_\n"} sort keys %configs;
}
sub VERSION_MESSAGE {
}

my %opts;
$Getopt::Std::STANDARD_HELP_VERSION = 1;
getopts('k', \%opts);

# If no config-name is provided, use all known configs.
# Otherwise, use the provided names only.
my @configs_to_test = sort keys %configs;
if ($#ARGV >= 0) {
    foreach my $conf_name ( @ARGV ) {
        if( ! exists $configs{$conf_name} ) {
            die "Unknown configuration: $conf_name\n";
        }
    }
    @configs_to_test = @ARGV;
}

-d 'library' && -d 'include' && -d 'tests' or die "Must be run from root\n";

my $config_h = 'include/mbedtls/mbedtls_config.h';

system( "cp $config_h $config_h.bak" ) and die;

my $failures = 0;

sub abort {
    warn $_[0];
    my $sig = $? & 127;
    system( "mv $config_h.bak $config_h" ) and warn "$config_h not restored\n";
    if ( $sig == POSIX::SIGINT )
    {
        warn "\nInterrupted during build\n";
        $SIG{'INT'} = 'DEFAULT';
        kill( $sig, $$ );
    }
    elsif ( $opts{k} )
    {
        die "failure";
    }
    else
    {
        # use an exit code between 1 and 124 for git bisect (die returns 255)
        exit 1;
    }
}

sub interrupt {
    warn "\nInterrupted during perl\n";
    die "interrupt";
}
$SIG{INT} = \&interrupt;

# Create a seedfile for configurations that enable MBEDTLS_ENTROPY_NV_SEED.
# For test purposes, this doesn't have to be cryptographically random.
if (!-e "tests/seedfile" || -s "tests/seedfile" < 64) {
    local *SEEDFILE;
    open SEEDFILE, ">tests/seedfile" or die;
    print SEEDFILE "*" x 64 or die;
    close SEEDFILE or die;
}

sub perform_test {
    my $conf = $_[0];
    my $conf_file = "configs/$conf";
    my $data = $_[1];
    my $test_with_psa = $_[2];

    if ( $test_with_psa )
    {
        $conf .= '+USE_PSA_CRYPTO';
    }

    system( "cp $config_h.bak $config_h" ) and die "fatal";
    system( "make clean" ) and die "fatal";

    print "\n******************************************\n";
    print "* Testing configuration: $conf\n";
    print "******************************************\n";

    $ENV{MBEDTLS_TEST_CONFIGURATION} = $conf;

    system( "cp $conf_file $config_h" )
        and abort "Failed to activate $conf\n";

    if ( $test_with_psa )
    {
        system( "scripts/config.py set MBEDTLS_PSA_CRYPTO_C" );
        system( "scripts/config.py set MBEDTLS_USE_PSA_CRYPTO" );
    }

    system( "CFLAGS='-Os -Werror -Wall -Wextra' make" ) and abort "Failed to build: $conf\n";
    system( "make test" ) and abort "Failed test suite: $conf\n";

    my $compat = $data->{'compat'};
    if( $compat )
    {
        print "\nrunning compat.sh $compat\n";
        system( "tests/compat.sh $compat" )
            and abort "Failed compat.sh: $conf\n";
    }
    else
    {
        print "\nskipping compat.sh\n";
    }

    my $opt = $data->{'opt'};
    if( $opt )
    {
        if( $data->{'opt_needs_debug'} )
        {
            print "\nrebuilding with debug traces for ssl-opt\n";
            system( "make clean" );
            system( "scripts/config.py set MBEDTLS_DEBUG_C" );
            system( "scripts/config.py set MBEDTLS_ERROR_C" );
            system( "CFLAGS='-Os -Werror -Wall -Wextra' make" ) and abort "Failed to build: $conf+DEBUG\n";
        }

        print "\nrunning ssl-opt.sh $opt\n";
        system( "tests/ssl-opt.sh $opt" )
            and abort "Failed ssl-opt.sh: $conf\n";
    }
    else
    {
        print "\nskipping ssl-opt.sh\n";
    }
}

sub wrap_test {
    eval {
        perform_test( @_ )
    };
    warn "\$\@ = \"$@\"\n";
    if ( $@ eq "failure" ) {
        ++$failures;
    } else {
        die $@;
    }
}

foreach my $conf ( @configs_to_test ) {
    my $test_with_psa = $configs{$conf}{'test_again_with_use_psa'};
    if ( $test_with_psa )
    {
        wrap_test( $conf, $configs{$conf}, $test_with_psa );
    }
    wrap_test( $conf, $configs{$conf}, 0 );
}

system( "mv $config_h.bak $config_h" ) and warn "$config_h not restored\n";
system( "make clean" );
exit !$failures;

File: tests/scripts/test_ref_configs.py
#!/usr/bin/env python3

"""Test reference configurations

For each reference configuration file in the configs directory, build the
configuration, run the test suites and, for some configurations, TLS tests.
If given one or more config name, only test these configurations.
"""

## Copyright The Mbed TLS Contributors
## SPDX-License-Identifier: Apache-2.0
##
## Licensed under the Apache License, Version 2.0 (the "License"); you may
## not use this file except in compliance with the License.
## You may obtain a copy of the License at
##
## http://www.apache.org/licenses/LICENSE-2.0
##
## Unless required by applicable law or agreed to in writing, software
## distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
## WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
## See the License for the specific language governing permissions and
## limitations under the License.

import argparse
import glob
import os
import shutil
import subprocess
import sys
import typing
from typing import Dict, Iterable, List

CONFIGS_DIR = 'configs'
LIVE_CONFIG = 'include/mbedtls/mbedtls_config.h'
BACKUP_CONFIG = LIVE_CONFIG + '.bak'
SEEDFILE = 'tests/seedfile'
SEEDFILE_SIZE = 64

class Spec:
    """Specification of additional testing for a configuration."""
    # In Python >=3.6, this could be a typing.NamedTuple. But Python 3.5
    # doesn't have NamedTuple with both types and default values.
    #pylint: disable=too-few-public-methods

    def __init__(self,
                 compat: Iterable[List[str]] = frozenset(),
                 opt: Iterable[List[str]] = frozenset(),
                 psa: bool = False):
        self.compat = compat
        self.opt = opt
        self.psa = psa

# Describe additional testing for some configurations. If a configuration is
# not listed here, this script only builds the library and runs the unit tests.
CONFIGS = {
    'config-ccm-psk-tls1_2.h': Spec(
        compat=[['-m', 'tls12', '-f', '^TLS-PSK-WITH-AES-...-CCM-8']],
        psa=True,
    ),
    'config-ccm-psk-dtls1_2.h': Spec(
        compat=[['-m', 'dtls12', '-f', '^TLS-PSK-WITH-AES-...-CCM-8']],
        psa=True,
    ),
    'config-suite-b.h': Spec(
        compat=[['-m', 'tls12', '-f', 'ECDHE-ECDSA.*AES.*GCM', '-p', 'mbedTLS']],
        psa=True,
    ),
    'config-symmetric-only.h': Spec(
        psa=False, # Uses PSA by default, no need to test it twice
    ),
    'config-thread.h': Spec(
        opt=[['-f', 'ECJPAKE.*nolog']],
        psa=True,
    ),
}

class TestOptions:
    """Runtime options for `test_configuration`."""
    #pylint: disable=too-few-public-methods

    keep_going = False #type: bool

def tweak_config(symbols_to_set: Iterable[str]) -> None:
    """Additionally set the given configuration options."""
    for symbol in symbols_to_set:
        subprocess.check_call(['scripts/config.py', 'set', symbol])

def test_configuration_variant(options: TestOptions,
                               config_name: str,
                               config_file: str,
                               spec: Spec = Spec(),
                               psa: bool = False) -> bool:
    """Test the specified configuration variant.

    Assume that `prepare_for_testing` has run.
    Running `final_cleanup` may be necessary afterwards.

    If `options.keep_going` is true, return True on success, False on failure.
    Otherwise raise an exception as soon as a failure occurs.
    """
    env = os.environ.copy()
    env['MBEDTLS_TEST_CONFIGURATION'] = config_name
    failures = [] #type: List[List[str]]
    def run(cmd: List[str], #pylint: disable=dangerous-default-value
            env: Dict[str, str] = env,
            **kwargs) -> bool:
        cp = subprocess.run(cmd,
                            env=env,
                            check=not options.keep_going,
                            **kwargs)
        if cp.returncode != 0:
            failures.append(cmd)
            return False
        return True
    # 1. Prepare
    print("""
******************************************
* Testing configuration: {}
******************************************
""".format(config_name))
    if not run(['make', 'clean']):
        return False
    shutil.copy(config_file, LIVE_CONFIG)
    if psa:
        tweak_config(['MBEDTLS_PSA_CRYPTO_C', 'MBEDTLS_USE_PSA_CRYPTO'])
    # 2. Build and run unit tests
    if not run(['make']):
        return False
    run(['make', 'test'])
    # 3. Run TLS tests
    for args in spec.compat:
        print("\nRunning compat.sh", *args)
        run(['tests/compat.sh'] + args)
    for args in spec.opt:
        print("\nRunning ssl-opt.sh", *args)
        run(['tests/ssl-opt.sh'] + args)
    # 4. In keep-going mode, report the failures
    return not failures

def all_configurations() -> List[str]:
    """List the available configurations."""
    return glob.glob(os.path.join(CONFIGS_DIR, '*.h'))

def test_configuration(options: TestOptions, config: str) -> int:
    """Test the given configuration.

    Assume that `prepare_for_testing` has run.
    Running `final_cleanup` may be necessary afterwards.

    If `options.keep_going` is true, return the number of failing variants
    (so 0 means success).
    Otherwise raise an exception as soon as a failure occurs.
    """
    if '/' in config:
        config_file = config
        config_name = os.path.basename(config)
    else:
        config_file = os.path.join(CONFIGS_DIR, config)
        config_name = config
    spec = CONFIGS.get(config_name, Spec())
    success = test_configuration_variant(options, config_name, config_file, spec)
    if spec.psa:
        success = success and \
            test_configuration_variant(options,
                                       config_name + '+PSA',
                                       config_file,
                                       spec,
                                       psa=True)
    return success

def test_configurations(options: TestOptions, configs: Iterable[str]) -> List[str]:
    """Test the specified configurations.

    If `options.keep_going` is true, return the list of failing
    configuration variants (so an empty list means success).
    Otherwise raise an exception as soon as a failure occurs.
    """
    failures = []
    for config in configs:
        if not test_configuration(options, config):
            failures.append(config)
    return failures

def prepare_for_testing() -> None:
    """Initial setup before running any tests."""
    shutil.copy(LIVE_CONFIG, BACKUP_CONFIG)
    # For configurations that enable MBEDTLS_ENTROPY_NV_SEED, ensure that
    # a seedfile is present.
    if not os.path.exists(SEEDFILE) or \
       os.stat(SEEDFILE).st_size < SEEDFILE_SIZE:
        with open(SEEDFILE, 'wb') as seedfile:
            seedfile.write(os.urandom(SEEDFILE_SIZE))

def final_cleanup() -> None:
    """Final cleanup after running all the tests."""
    shutil.move(BACKUP_CONFIG, LIVE_CONFIG)
    subprocess.check_call(['make', 'clean'])

def run_tests(options: TestOptions, configs: List[str]) -> List[str]:
    """Test the specified configurations, then clean up.

    If `options.keep_going` is true, return the number of failing variants
    (so 0 means success).
    Otherwise raise an exception as soon as a failure occurs.
    """
    prepare_for_testing()
    try:
        return test_configurations(options, configs)
    finally:
        final_cleanup()

def report_tests(options: TestOptions, configs: List[str]) -> int:
    """Test the specified configurations, then clean up.

    If `options.keep_going` is true, return the number of failing variants
    (so 0 means success).
    Otherwise raise an exception as soon as a failure occurs.
    """
    failures = run_tests(options, configs)
    if failures:
        # The number of failures counts configuration variants:
        # foo.h and foo.h+PSA together count as 2.
        print('{}: FAILED: {}'.format(sys.argv[0], ' '.join(failures)))
    else:
        # The number of passes counts configurations:
        # foo.h and foo.h+PSA together count as 1.
        print('{}: {} PASSED'.format(sys.argv[0], len(configs)))
    return len(failures)

def main() -> None:
    """Command line entry point."""
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--keep-going', '-k',
                        action='store_true', default=False,
                        help='Keep going after an error')
    parser.add_argument('--stop', '-S',
                        action='store_false', dest='keep_going',
                        help='Stop on error (opposite of --keep-going)')
    parser.add_argument('configs', nargs='*', metavar='CONFIGS',
                        help=('Configurations to test'
                              ' (files; relative to configs/ if no slash;'
                              ' default: all in configs/)'))
    options = parser.parse_args()
    configs = options.configs if options.configs else all_configurations()
    sys.exit(report_tests(typing.cast(TestOptions, options), configs))

if __name__ == '__main__':
    main()
